crates/ark/src/variables/variable.rs

```diff
@@ -918,132 +931,216 @@ impl PositronVariable {
+            let name = match parsed_access_key[3] {
+                "" => RObject::from(r_null()), // Empty string, means a `NULL` name
+                nm => {
+                    if nm.len() == name_len {
+                        RObject::from(nm)
+                    } else {
+                        // Name has been truncated, we pass it as `NULL`
+                        RObject::from(r_null())
                     }
                 },
+            };
 
-                EnvironmentVariableNode::VectorElement { .. } => {
-                    return Err(harp::error::Error::InspectError { path: path.clone() });
-                },
+            ArkGenerics::VariableGetChildAt.try_dispatch::<RObject>(object.sexp, vec![
+                RArgument::new("index", RObject::from(index + 1)), // Index is 0-based, so we convert to 1-based for R.
+                RArgument::new("name", RObject::from(name)),
+            ])
+        });
 
-                EnvironmentVariableNode::Matrixcolumn { object, index } => unsafe {
-                    let dim = IntegerVector::new(Rf_getAttrib(*object, R_DimSymbol))?;
-                    let n_row = dim.get_unchecked(0).unwrap() as isize;
+        match result {
+            Err(err) => {
+                // It's not safe to apply default methods in this case, because we rely on custom
+                // access keys, which could indicate the access index depending on the node implementation.
+                // See for instance, how it's used to index lists and atomic vectors.
+                return Err(harp::Error::Anyhow(err));
+            },
+            Ok(None) => {
+                // The object doesn't have a custom get_child_at method. We apply
+                // the default built-in methods.
+            },
+            Ok(Some(child)) => return Ok(EnvironmentVariableNode::Concrete { object: child }),
+        };
+
+        // For S4 objects, we acess child nodes using R_do_slot.
+        if object.is_s4() {
+            let name = unsafe { r_symbol!(access_key) };
+            let child: RObject =
+                harp::try_catch(|| unsafe { R_do_slot(object.sexp, name) }.into())?;
+            return Ok(EnvironmentVariableNode::Concrete { object: child });
+        }
 
-                    // TODO: use ? here, but this does not return a crate::error::Error, so
-                    //       maybe use anyhow here instead ?
-                    let row_index = path_element.parse::<isize>().unwrap();
+        // R6 objects may be accessed with special elements called <methods> and <private>.
+        // For them, we'll have to build the next node artifically.
+        if r_inherits(object.sexp, "R6") && access_key.starts_with("<") {
+            return Ok(EnvironmentVariableNode::R6Node {
+                object,
+                name: access_key.clone(),
+            });
+        }
 
-                    EnvironmentVariableNode::VectorElement {
+        match r_typeof(*object) {
```

Comment:
While you're in there, could you please change all `*obj` to the clearer `obj.sexp`?

Reply:
Done in https://github.com/posit-dev/ark/pull/561/commits/4c9b962340639f3d07ac7628ddf717ba7603c2d1